pallene/C.lua

```diff
@@ -37,15 +37,21 @@ function C.float(n)
     -- 17 decimal digits should be able to accurately represent any IEE-754
     -- double-precision floating point number except for NaN and infinities.
     -- The HUGE_VAL macro is a part of math.h. For more info, please see the
-    -- quotefloat function in lstrlib.c and https://stackoverflow.com/a/21162120
+    -- quotefloat function in lstrlib.c, tostringbuff in lobject.c, and
+    -- https://stackoverflow.com/a/21162120
     if n == math.huge then
         return "HUGE_VAL"
     elseif n == -math.huge then
         return "-HUGE_VAL"
     elseif n ~= n then
         error("NaN cannot be round-tripped")
     else
-        return string.format("%.17g", n)
+        local s = string.format("%.17g", n)
+        if s:match("^%-?[0-9]+$") then
```

Comment:
This seems overly-complicated. Couldn't we use `%e`?

Reply:
Adding the ".0" like this is the same idea that Lua uses in its `tostring` function.

The problem with %e is that, like %f, it always prints 17 digits after the decimal dot. For example, 0.0 becomes 0.00000000000000000e+00

Reply:
Well, if you are worried readability of the generated code this makes sense. Anyway, I would rewrite the code as the following:
```
if isinteger(n) then
   string.format("%d.0", n)
else
  string.format("%.17g", n)
end
```
`isinteger` could be something like `n == math.ceil(n)`

Reply:
I tested this version and unfortunately it does not always work. For large numbers such as 1e19, isinteger can be true but the number is too large to be converted to integer. In Lua, trying to print them with %d gives an error, while in C it leads to integer overflow.